File: tinylog-jboss/src/org/pmw/tinylog/Logger.java
/*
 * Copyright 2016 Martin Winandy
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package org.pmw.tinylog;

import org.pmw.tinylog.runtime.RuntimeDialect;

/**
 * Static class to create log entries.
 */
public final class Logger {

	private static final int DEPTH_OF_STACK_TRACE = 3;
	private static final String FQCN = Logger.class.getName();

	private static final RuntimeDialect runtime = EnvironmentHelper.getRuntimeDialect();

	private Logger() {
	}

	/**
	 * Get the current global severity level.
	 *
	 * @return Global severity level
	 */
	public static Level getLevel() {
		return getLevel(org.jboss.logging.Logger.getLogger(""));
	}

	/**
	 * Get the current severity level for a specific package.
	 *
	 * @param packageObject
	 *            Package
	 *
	 * @return Severity level for the package
	 */
	public static Level getLevel(final Package packageObject) {
		return getLevel(org.jboss.logging.Logger.getLogger(packageObject.getName()));
	}

	/**
	 * Get the current severity level for a specific class.
	 *
	 * @param classObject
	 *            Name of the class
	 *
	 * @return Severity level for the class
	 */
	public static Level getLevel(final Class<?> classObject) {
		return getLevel(org.jboss.logging.Logger.getLogger(classObject));
	}

	/**
	 * Get the current severity level for a specific package or class.
	 *
	 * @param packageOrClass
	 *            Name of the package respectively class
	 *
	 * @return severity level for the package respectively class
	 */
	public static Level getLevel(final String packageOrClass) {
		return getLevel(org.jboss.logging.Logger.getLogger(packageOrClass));
	}

	/**
	 * Create a trace log entry.
	 *
	 * @param obj
	 *            The result of the <code>toString()</code> method will be logged
	 */
	public static void trace(final Object obj) {
		getJBossLogger().trace(FQCN, obj, null);
	}

	/**
	 * Create a trace log entry.
	 *
	 * @param message
	 *            Text message to log
	 */
	public static void trace(final String message) {
		getJBossLogger().trace(FQCN, message, null);
	}

	/**
	 * Create a trace log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void trace(final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.TRACE)) {
			logger.trace(FQCN, MessageFormatter.format(message, arguments), null);
		}
	}

	/**
	 * Create a trace log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param exception
	 *            Exception to log
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void trace(final Throwable exception, final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.TRACE)) {
			logger.trace(FQCN, MessageFormatter.format(message, arguments), exception);
		}
	}

	/**
	 * Create a trace log entry.
	 *
	 * @param exception
	 *            Exception to log
	 */
	public static void trace(final Throwable exception) {
		getJBossLogger().trace(FQCN, exception, exception);
	}

	/**
	 * Create a debug log entry.
	 *
	 * @param obj
	 *            The result of the <code>toString()</code> method will be logged
	 */
	public static void debug(final Object obj) {
		getJBossLogger().debug(FQCN, obj, null);
	}

	/**
	 * Create a debug log entry.
	 *
	 * @param message
	 *            Text message to log
	 */
	public static void debug(final String message) {
		getJBossLogger().debug(FQCN, message, null);
	}

	/**
	 * Create a debug log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void debug(final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.DEBUG)) {
			logger.debug(FQCN, MessageFormatter.format(message, arguments), null);
		}
	}

	/**
	 * Create a debug log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param exception
	 *            Exception to log
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void debug(final Throwable exception, final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.DEBUG)) {
			logger.debug(FQCN, MessageFormatter.format(message, arguments), exception);
		}
	}

	/**
	 * Create a debug log entry.
	 *
	 * @param exception
	 *            Exception to log
	 */
	public static void debug(final Throwable exception) {
		getJBossLogger().debug(FQCN, exception, exception);
	}

	/**
	 * Create an info log entry.
	 *
	 * @param obj
	 *            The result of the <code>toString()</code> method will be logged
	 */
	public static void info(final Object obj) {
		getJBossLogger().info(FQCN, obj, null);
	}

	/**
	 * Create an info log entry.
	 *
	 * @param message
	 *            Text message to log
	 */
	public static void info(final String message) {
		getJBossLogger().info(FQCN, message, null);
	}

	/**
	 * Create an info log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void info(final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.INFO)) {
			logger.info(FQCN, MessageFormatter.format(message, arguments), null);
		}
	}

	/**
	 * Create an info log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param exception
	 *            Exception to log
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void info(final Throwable exception, final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.INFO)) {
			logger.info(FQCN, MessageFormatter.format(message, arguments), exception);
		}
	}

	/**
	 * Create an info log entry.
	 *
	 * @param exception
	 *            Exception to log
	 */
	public static void info(final Throwable exception) {
		getJBossLogger().info(FQCN, exception, exception);
	}

	/**
	 * Create a warning log entry.
	 *
	 * @param obj
	 *            The result of the <code>toString()</code> method will be logged
	 */
	public static void warn(final Object obj) {
		getJBossLogger().warn(FQCN, obj, null);
	}

	/**
	 * Create a warning log entry.
	 *
	 * @param message
	 *            Text message to log
	 */
	public static void warn(final String message) {
		getJBossLogger().warn(FQCN, message, null);
	}

	/**
	 * Create a warning log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void warn(final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.WARN)) {
			logger.warn(FQCN, MessageFormatter.format(message, arguments), null);
		}
	}

	/**
	 * Create a warning log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param exception
	 *            Exception to log
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void warn(final Throwable exception, final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.WARN)) {
			logger.warn(FQCN, MessageFormatter.format(message, arguments), exception);
		}
	}

	/**
	 * Create a warning log entry.
	 *
	 * @param exception
	 *            Exception to log
	 */
	public static void warn(final Throwable exception) {
		getJBossLogger().warn(FQCN, exception, exception);
	}

	/**
	 * Create an error log entry.
	 *
	 * @param obj
	 *            The result of the <code>toString()</code> method will be logged
	 */
	public static void error(final Object obj) {
		getJBossLogger().error(FQCN, obj, null);
	}

	/**
	 * Create an error log entry.
	 *
	 * @param message
	 *            Text message to log
	 */
	public static void error(final String message) {
		getJBossLogger().error(FQCN, message, null);
	}

	/**
	 * Create an error log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void error(final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.ERROR)) {
			logger.error(FQCN, MessageFormatter.format(message, arguments), null);
		}
	}

	/**
	 * Create an error log entry. "{}" placeholders will be replaced by the given arguments.
	 *
	 * @param exception
	 *            Exception to log
	 * @param message
	 *            Formated text for the log entry
	 * @param arguments
	 *            Arguments for the text message
	 */
	public static void error(final Throwable exception, final String message, final Object... arguments) {
		org.jboss.logging.Logger logger = getJBossLogger();
		if (logger.isEnabled(org.jboss.logging.Logger.Level.ERROR)) {
			logger.error(FQCN, MessageFormatter.format(message, arguments), exception);
		}
	}

	/**
	 * Create an error log entry.
	 *
	 * @param exception
	 *            Exception to log
	 */
	public static void error(final Throwable exception) {
		getJBossLogger().error(FQCN, exception, exception);
	}

	private static org.jboss.logging.Logger getJBossLogger() {
		return org.jboss.logging.Logger.getLogger(runtime.getClassName(DEPTH_OF_STACK_TRACE));
	}

	private static Level getLevel(final org.jboss.logging.Logger logger) {
		if (logger.isEnabled(org.jboss.logging.Logger.Level.TRACE)) {
			return Level.TRACE;
		} else if (logger.isEnabled(org.jboss.logging.Logger.Level.DEBUG)) {
			return Level.DEBUG;
		} else if (logger.isEnabled(org.jboss.logging.Logger.Level.INFO)) {
			return Level.INFO;
		} else if (logger.isEnabled(org.jboss.logging.Logger.Level.WARN)) {
			return Level.WARNING;
		} else if (logger.isEnabled(org.jboss.logging.Logger.Level.ERROR)) {
			return Level.ERROR;
		} else {
			return Level.OFF;
		}
	}

}

File: tinylog-api-scala/src/main/scala/org/tinylog/scala/Logger.scala
/*
 * Copyright 2019 Martin Winandy
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package org.tinylog.scala

import scala.language.experimental.macros

/**
	* Static logger for issuing log entries.
	*/
object Logger {

	/**
		* Checks whether log entries at [[org.tinylog.Level#TRACE]] will be output.
		*
		* @return `true` if [[org.tinylog.Level#TRACE]] level is enabled, `false` if disabled
		*/
	def isTraceEnabled(): Boolean = macro LoggerMacro.isTraceEnabled

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* @param message
		* Any object with a meaningful `toString()` method
		*/
	def trace(message: Any): Unit = macro LoggerMacro.tracePlainMessage

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* Strings with embedded variables will be evaluated lazy by a macro.
		*
		* @param message
		* Text message to log
		*/
	def trace(message: String): Unit = macro LoggerMacro.tracePlainMessage

	/**
		* Logs a lazy message at [[org.tinylog.Level#TRACE]]. The message will be only evaluated if the log entry is
		* really output.
		*
		* @param message
		* Function that produces the message
		*/
	def trace(message: () => String): Unit = macro LoggerMacro.traceLazyMessage

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced by given
		* arguments.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def trace(message: String, arguments: Any*): Unit = macro LoggerMacro.traceMessageWithPlainArguments

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]] level. "{}" placeholders will be replaced by given lazy
		* arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def trace(message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.traceMessageWithLazyArguments

	/**
		* Logs an exception at [[org.tinylog.Level#TRACE]].
		*
		* @param exception
		* Caught exception or any other throwable to log
		*/
	def trace(exception: Throwable): Unit = macro LoggerMacro.traceException

	/**
		* Logs an exception with a custom message at [[org.tinylog.Level#TRACE]].
		*
		* Messages with embedded variables will be evaluated lazy by a macro.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Text message to log
		*/
	def trace(exception: Throwable, message: String): Unit = macro LoggerMacro.traceExceptionWithPlainMessage

	/**
		* Logs an exception with a custom lazy message at [[org.tinylog.Level#TRACE]]. The message will be only
		* evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Function that produces the message
		*/
	def trace(exception: Throwable, message: () => String): Unit = macro LoggerMacro.traceExceptionWithLazyMessage

	/**
		* Logs an exception with a formatted custom message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be
		* replaced by given arguments.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def trace(exception: Throwable, message: String, arguments: Any*): Unit = macro LoggerMacro.traceExceptionWithMessageWithPlainArguments

	/**
		* Logs an exception with a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced
		* by given lazy arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def trace(exception: Throwable, message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.traceExceptionWithMessageWithLazyArguments

	/**
		* Checks whether log entries at [[org.tinylog.Level#TRACE]] will be output.
		*
		* @return `true` if [[org.tinylog.Level#TRACE]] level is enabled, `false` if disabled
		*/
	def isDebugEnabled(): Boolean = macro LoggerMacro.isDebugEnabled

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* @param message
		* Any object with a meaningful `toString()` method
		*/
	def debug(message: Any): Unit = macro LoggerMacro.debugPlainMessage

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* Strings with embedded variables will be evaluated lazy by a macro.
		*
		* @param message
		* Text message to log
		*/
	def debug(message: String): Unit = macro LoggerMacro.debugPlainMessage

	/**
		* Logs a lazy message at [[org.tinylog.Level#TRACE]]. The message will be only evaluated if the log entry is
		* really output.
		*
		* @param message
		* Function that produces the message
		*/
	def debug(message: () => String): Unit = macro LoggerMacro.debugLazyMessage

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced by given
		* arguments.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def debug(message: String, arguments: Any*): Unit = macro LoggerMacro.debugMessageWithPlainArguments

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]] level. "{}" placeholders will be replaced by given lazy
		* arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def debug(message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.debugMessageWithLazyArguments

	/**
		* Logs an exception at [[org.tinylog.Level#TRACE]].
		*
		* @param exception
		* Caught exception or any other throwable to log
		*/
	def debug(exception: Throwable): Unit = macro LoggerMacro.debugException

	/**
		* Logs an exception with a custom message at [[org.tinylog.Level#TRACE]].
		*
		* Messages with embedded variables will be evaluated lazy by a macro.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Text message to log
		*/
	def debug(exception: Throwable, message: String): Unit = macro LoggerMacro.debugExceptionWithPlainMessage

	/**
		* Logs an exception with a custom lazy message at [[org.tinylog.Level#TRACE]]. The message will be only
		* evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Function that produces the message
		*/
	def debug(exception: Throwable, message: () => String): Unit = macro LoggerMacro.debugExceptionWithLazyMessage

	/**
		* Logs an exception with a formatted custom message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be
		* replaced by given arguments.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def debug(exception: Throwable, message: String, arguments: Any*): Unit = macro LoggerMacro.debugExceptionWithMessageWithPlainArguments

	/**
		* Logs an exception with a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced
		* by given lazy arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def debug(exception: Throwable, message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.debugExceptionWithMessageWithLazyArguments

	/**
		* Checks whether log entries at [[org.tinylog.Level#TRACE]] will be output.
		*
		* @return `true` if [[org.tinylog.Level#TRACE]] level is enabled, `false` if disabled
		*/
	def isInfoEnabled(): Boolean = macro LoggerMacro.isInfoEnabled

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* @param message
		* Any object with a meaningful `toString()` method
		*/
	def info(message: Any): Unit = macro LoggerMacro.infoPlainMessage

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* Strings with embedded variables will be evaluated lazy by a macro.
		*
		* @param message
		* Text message to log
		*/
	def info(message: String): Unit = macro LoggerMacro.infoPlainMessage

	/**
		* Logs a lazy message at [[org.tinylog.Level#TRACE]]. The message will be only evaluated if the log entry is
		* really output.
		*
		* @param message
		* Function that produces the message
		*/
	def info(message: () => String): Unit = macro LoggerMacro.infoLazyMessage

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced by given
		* arguments.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def info(message: String, arguments: Any*): Unit = macro LoggerMacro.infoMessageWithPlainArguments

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]] level. "{}" placeholders will be replaced by given lazy
		* arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def info(message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.infoMessageWithLazyArguments

	/**
		* Logs an exception at [[org.tinylog.Level#TRACE]].
		*
		* @param exception
		* Caught exception or any other throwable to log
		*/
	def info(exception: Throwable): Unit = macro LoggerMacro.infoException

	/**
		* Logs an exception with a custom message at [[org.tinylog.Level#TRACE]].
		*
		* Messages with embedded variables will be evaluated lazy by a macro.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Text message to log
		*/
	def info(exception: Throwable, message: String): Unit = macro LoggerMacro.infoExceptionWithPlainMessage

	/**
		* Logs an exception with a custom lazy message at [[org.tinylog.Level#TRACE]]. The message will be only
		* evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Function that produces the message
		*/
	def info(exception: Throwable, message: () => String): Unit = macro LoggerMacro.infoExceptionWithLazyMessage

	/**
		* Logs an exception with a formatted custom message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be
		* replaced by given arguments.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def info(exception: Throwable, message: String, arguments: Any*): Unit = macro LoggerMacro.infoExceptionWithMessageWithPlainArguments

	/**
		* Logs an exception with a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced
		* by given lazy arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def info(exception: Throwable, message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.infoExceptionWithMessageWithLazyArguments

	/**
		* Checks whether log entries at [[org.tinylog.Level#TRACE]] will be output.
		*
		* @return `true` if [[org.tinylog.Level#TRACE]] level is enabled, `false` if disabled
		*/
	def isWarnEnabled(): Boolean = macro LoggerMacro.isWarnEnabled

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* @param message
		* Any object with a meaningful `toString()` method
		*/
	def warn(message: Any): Unit = macro LoggerMacro.warnPlainMessage

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* Strings with embedded variables will be evaluated lazy by a macro.
		*
		* @param message
		* Text message to log
		*/
	def warn(message: String): Unit = macro LoggerMacro.warnPlainMessage

	/**
		* Logs a lazy message at [[org.tinylog.Level#TRACE]]. The message will be only evaluated if the log entry is
		* really output.
		*
		* @param message
		* Function that produces the message
		*/
	def warn(message: () => String): Unit = macro LoggerMacro.warnLazyMessage

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced by given
		* arguments.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def warn(message: String, arguments: Any*): Unit = macro LoggerMacro.warnMessageWithPlainArguments

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]] level. "{}" placeholders will be replaced by given lazy
		* arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def warn(message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.warnMessageWithLazyArguments

	/**
		* Logs an exception at [[org.tinylog.Level#TRACE]].
		*
		* @param exception
		* Caught exception or any other throwable to log
		*/
	def warn(exception: Throwable): Unit = macro LoggerMacro.warnException

	/**
		* Logs an exception with a custom message at [[org.tinylog.Level#TRACE]].
		*
		* Messages with embedded variables will be evaluated lazy by a macro.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Text message to log
		*/
	def warn(exception: Throwable, message: String): Unit = macro LoggerMacro.warnExceptionWithPlainMessage

	/**
		* Logs an exception with a custom lazy message at [[org.tinylog.Level#TRACE]]. The message will be only
		* evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Function that produces the message
		*/
	def warn(exception: Throwable, message: () => String): Unit = macro LoggerMacro.warnExceptionWithLazyMessage

	/**
		* Logs an exception with a formatted custom message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be
		* replaced by given arguments.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def warn(exception: Throwable, message: String, arguments: Any*): Unit = macro LoggerMacro.warnExceptionWithMessageWithPlainArguments

	/**
		* Logs an exception with a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced
		* by given lazy arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def warn(exception: Throwable, message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.warnExceptionWithMessageWithLazyArguments

	/**
		* Checks whether log entries at [[org.tinylog.Level#TRACE]] will be output.
		*
		* @return `true` if [[org.tinylog.Level#TRACE]] level is enabled, `false` if disabled
		*/
	def isErrorEnabled(): Boolean = macro LoggerMacro.isErrorEnabled

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* @param message
		* Any object with a meaningful `toString()` method
		*/
	def error(message: Any): Unit = macro LoggerMacro.errorPlainMessage

	/**
		* Logs a message at [[org.tinylog.Level#TRACE]].
		*
		* Strings with embedded variables will be evaluated lazy by a macro.
		*
		* @param message
		* Text message to log
		*/
	def error(message: String): Unit = macro LoggerMacro.errorPlainMessage

	/**
		* Logs a lazy message at [[org.tinylog.Level#TRACE]]. The message will be only evaluated if the log entry is
		* really output.
		*
		* @param message
		* Function that produces the message
		*/
	def error(message: () => String): Unit = macro LoggerMacro.errorLazyMessage

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced by given
		* arguments.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def error(message: String, arguments: Any*): Unit = macro LoggerMacro.errorMessageWithPlainArguments

	/**
		* Logs a formatted message at [[org.tinylog.Level#TRACE]] level. "{}" placeholders will be replaced by given lazy
		* arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def error(message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.errorMessageWithLazyArguments

	/**
		* Logs an exception at [[org.tinylog.Level#TRACE]].
		*
		* @param exception
		* Caught exception or any other throwable to log
		*/
	def error(exception: Throwable): Unit = macro LoggerMacro.errorException

	/**
		* Logs an exception with a custom message at [[org.tinylog.Level#TRACE]].
		*
		* Messages with embedded variables will be evaluated lazy by a macro.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Text message to log
		*/
	def error(exception: Throwable, message: String): Unit = macro LoggerMacro.errorExceptionWithPlainMessage

	/**
		* Logs an exception with a custom lazy message at [[org.tinylog.Level#TRACE]]. The message will be only
		* evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Function that produces the message
		*/
	def error(exception: Throwable, message: () => String): Unit = macro LoggerMacro.errorExceptionWithLazyMessage

	/**
		* Logs an exception with a formatted custom message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be
		* replaced by given arguments.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Arguments for formatted text message
		*/
	def error(exception: Throwable, message: String, arguments: Any*): Unit = macro LoggerMacro.errorExceptionWithMessageWithPlainArguments

	/**
		* Logs an exception with a formatted message at [[org.tinylog.Level#TRACE]]. "{}" placeholders will be replaced
		* by given lazy arguments. The arguments will be only evaluated if the log entry is really output.
		*
		* @param exception
		* Caught exception or any other throwable to log
		* @param message
		* Formatted text message to log
		* @param arguments
		* Functions that produce the arguments for formatted text message
		*/
	def error(exception: Throwable, message: String, arguments: (() => Any)*): Unit = macro LoggerMacro.errorExceptionWithMessageWithLazyArguments

}
